zerver/lib/mention.py

```diff
@@ -162,15 +177,59 @@ def get_stream_name_map(self, stream_names: Set[str]) -> Dict[str, int]:
                 .values(
                     "id",
                     "name",
+                    "recipient_id",
+                    "history_public_to_subscribers",
                 )
             )
 
             for row in rows:
                 self.stream_cache[row["name"]] = row["id"]
+                self.recipient_cache[row["name"]] = ChannelInfo(
+                    row["recipient_id"], row["history_public_to_subscribers"]
+                )
                 result[row["name"]] = row["id"]
 
         return result
 
+    def get_topic_info_map(
+        self, channel_topic: Set[ChannelTopicInfo], message_sender: Optional[UserProfile]
+    ) -> Dict[ChannelTopicInfo, int]:
+        if not channel_topic:
+            return {}
+
+        result: Dict[ChannelTopicInfo, int] = {}
+        unseen_channel_topic: List[ChannelTopicInfo] = []
+
+        for channel_topic_object in channel_topic:
+            if channel_topic_object in self.topic_cache:
+                result[channel_topic_object] = self.topic_cache[channel_topic_object]
+            else:
+                unseen_channel_topic.append(channel_topic_object)
+
+        for channel_topic_object in unseen_channel_topic:
+            channel_info = self.recipient_cache.get(channel_topic_object.channel_name)
+
+            assert channel_info is not None
+            recipient_id = channel_info.recipient_id
+            topic_name = channel_topic_object.topic_name
+            history_public_to_subscribers = channel_info.history_public_to_subscribers
```

Comment:
Where does this logic check that the sending user has access to the channel at all? I think that's a prerequisite for the rest of this behavior to run. `can_access_stream_history` may be most useful? You'll want to be careful to minimize database queries.

Reply:
In the line right below, we basically try to fetch the last message in the topic, wherein if the user does not have access to the message (no `UserMessage` row and `not history_public_to_subscribers`), then `with` operator won't be added, and we'll be resorting back to the old topic links.